File: py/ch02/E02E_greedy_motif_search_pseudocounts.py
from utils import text, read_lines, score, bases
from ch02.E02C_profile_most_probable_kmer import MostProbableKmer


def GreedyMotifSearchWithPseudocounts(dna, k, t):
    best_motifs = [string[:k] for string in dna]

    for kmer in [text(dna[0], i, k) for i in range(len(dna[0]) - k + 1)]:
        motifs = [kmer]

        for i in range(1, t):
            _motifs = motifs[:i]
            matrix = []

            for base in bases:
                mat = []

                for j in range(k):
                    mm = [m[j] for m in _motifs]
                    mat.append((mm.count(base) + 1) / (len(mm) + 4))

                matrix.append(mat)

            motifs.append(MostProbableKmer(dna[i], k, matrix=matrix))

        if score(motifs) < score(best_motifs):
            best_motifs = motifs

    return best_motifs


if __name__ == "__main__":
    _dna, _k = read_lines(end_with='int')
    _t = int(input("t: "))

    for _word in GreedyMotifSearchWithPseudocounts(_dna, _k, _t):
        print(_word)

File: cpp/ch02/E02E_greedy_motif_search_pseudocounts.cpp
//
// Created by Matej Hrnčiar on 28/07/2022.
//

#include "E02E_greedy_motif_search_pseudocounts.h"

vector<string> GreedyMotifSearchPseudocounts(const vector<string> &dna, int k, int t) {
    vector<string> bestMotifs, kmers;

    for (auto &str : dna) {
        bestMotifs.push_back(str.substr(0, k));
    }

    for (int i = 0; i < dna[0].size() - k + 1; i++) {
        kmers.push_back(Text(dna[0], i, k));
    }

    for (auto &kmer : kmers) {
        vector<string> motifs;
        motifs.push_back(kmer);

        for (int i = 1; i < t; i++) {
            vector<string> _motifs;
            vector< vector<float> > matrix;

            for (int j = 0; j < i; j++) {
                _motifs.push_back(motifs[j]);
            }

            for (auto &base : bases) {
                vector<float> mat;

                for (int j = 0; j < k; j++) {
                    vector<string> mm;
                    for (auto &m : _motifs) {
                        mm.push_back(m.substr(j, 1));
                    }

                    float total = 0;
                    for (auto &temp : mm) {
                        if (base == temp)
                            total++;
                    }

                    mat.push_back((total + 1) / (float) (_motifs.size() + 4));
                }

                matrix.push_back(mat);
            }

            motifs.push_back(MostProbableKmer(dna[i], k, matrix));
        }

        if (Score(motifs) < Score(bestMotifs)) {
            bestMotifs = motifs;
        }
    }

    return bestMotifs;
}

int main() {
    vector<string> _dna;
    string _temp;
    int _k, _t;

    cout << "k: ";
    cin >> _k;
    cout << "t: ";
    cin >> _t;

    _dna = ReadLines();

    auto _motifs = GreedyMotifSearchPseudocounts(_dna, _k, _t);

    for (const auto &word : _motifs) {
        cout << word << endl;
    }
}
